[HW4] Patterns-GoF

AlwaysSayNo opened the pull request.

Commit:
[Refactoring] (Updated routing prefix)

- Added v1 prefix to all routes.

Commit:
[New provider] (Added new currency provider)

- Added new provider cdn.jsdelivr for currency.
- Updated old provider name.

Commit:
[GoF] (Implemented chain of responsibility)

- Implemented chain of responsibility: Private Bank API -> CDN.JSDELIVR.
- Updated currency service class to accept chain.

Commit:
[Refactoring] (Refactored EmailService)

- After CurrencyService changes updated EmailService to use it correctly.
- Updated email_template.html since we don't have sale/buy information - only one number.

Commit:
[Refactoring] (Refactored services)

- Refactored CurrencyService to return cache currency.
- Refactored UserService to add alias for my errors.

Commit:
[New provider] (Added and implemented new currency provider)

- Added new provider bank.gov.ua for currency.
- Implemented chain of responsibility: Private Bank API -> BANK.GOV.UA -> CDN.JSDELIVR.

Commit:
[Refactoring] (Folder restructuring; Separating common currency logic)

- Refactored folder structure for external currency APIs;
- Moved common currency API logic into abstract currency client which is used in other implementations internally.

Commit:
[Repository] (Implemented User Repository)

- Implemented pattern Repository for entity User.

Commit:
[Refactoring] (Variables/structures renaming)

- Renamed variables, structures after service files refactoring.

Commit:
Merge branch 'main' into hw4-patterns-gof

Commit:
[Refactoring] (Naming refactoring)

andriihar commented on internal/external/api/currency/abstract/abstract.go:
```diff
@@ -0,0 +1,71 @@
+
+// ProcessCurrencyResponseDTO returns information about currency rate.
+func (s *Client) ProcessCurrencyResponseDTO(responseDTO *dto.CurrencyResponseDTO, err error) (*dto.CurrencyResponseDTO, error) {
+	if err == nil {
+		log.Printf("Success response from %s: %v\n", s.ProviderName, *responseDTO)
+		return responseDTO, nil
+	} else if s.Next == nil {
+		return nil, fmt.Errorf("end of the currency rater chain: %w", err)
+	} else {
+		log.Printf("Error while calling %s: %v", s.ProviderName, err)
+	}
```
Can we use here early return?

https://medium.com/swlh/return-early-pattern-3d18a41bba8

AlwaysSayNo replied:
Removed one statement. Thank you for a good article.

andriihar commented on internal/external/api/currency/abstract/abstract.go:
```diff
@@ -0,0 +1,71 @@
+	return nil
+}
+
+func (s *Client) SetNext(next CurrencyRater) {
```
Іs it used?

AlwaysSayNo replied:
Thank your sharp eyes, I reused this code in the lower currency implimentations.

andriihar commented on internal/external/api/currency/abstract/abstract.go:
```diff
@@ -0,0 +1,71 @@
+
+// ProcessCurrencyResponseDTO returns information about currency rate.
+func (s *Client) ProcessCurrencyResponseDTO(responseDTO *dto.CurrencyResponseDTO, err error) (*dto.CurrencyResponseDTO, error) {
+	if err == nil {
```
Also, why it`s here?

AlwaysSayNo replied:
Sorry, didn't get your question?

andriihar replied:
I mean, why are we error checking here, maybe it should be a previous level?

AlwaysSayNo replied:
Ah, I see. Because all raters should have a common logic of fetching data from the next API in case if an error happend with the currenct rater. So, I moved this "common" thing into an abstract rater.

andriihar commented on internal/service/currency/service.go:
```diff
@@ -0,0 +1,72 @@
+// GetCurrencyRate returns short information about currency rate.
+func (s *Service) GetCurrencyRate() (dto.CurrencyResponseDTO, error) {
+	currencyDTO, err := s.getCurrencyDTO()
+	return currencyDTO.CurrencyResponseDTO, err
```
<img width="937" alt="image" src="https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-AlwaysSayNo/assets/25195930/e2660779-cf78-4fc6-8e94-461f32cc6e51">

AlwaysSayNo replied:
Fixed

andriihar commented on internal/service/user/service.go:
```diff
@@ -0,0 +1,62 @@
+
+//todo update docs
+
+type ServiceInterface interface {
```
Does it usage?

AlwaysSayNo replied:
Removed

andriihar commented on internal/service/user/service.go:
```diff
@@ -0,0 +1,62 @@
+}
+
+type Service struct {
+	userRepository *user.Repository
```
```suggestion
	userRepository ServiceInterface
```

?

AlwaysSayNo replied:
Changed

Commit:
[Refactoring, Tests] (Refactored some files; Wrote new tests; Fixed old tests)

- Fixed old tests, so they can work with new logic.
- Fixed mocks logic so it is more clear.
- Refactored naming in some files.
- Added user unit tests.

Commit:
[Test] (Added build step)

Commit:
[Refactoring] (Refactoring after PR)

Commit:
[Refactoring] (Refactored linet workflow)
